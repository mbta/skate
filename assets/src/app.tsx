// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
// tslint:disable-next-line
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "@babel/polyfill"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "leaflet/dist/leaflet.css"
import "phoenix_html"
import * as React from "react"
import ReactDOM from "react-dom"
import AppStateWrapper from "./components/appStateWrapper"

if (window.FS && window.username) {
  window.FS.identify(window.username, { displayName: window.username })
}

// Import local files
//
// Local files can be imported directly using relative paths, for example:
// import socket from "./socket"

ReactDOM.render(<AppStateWrapper />, document.getElementById("app"))
