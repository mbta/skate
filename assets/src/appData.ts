const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
