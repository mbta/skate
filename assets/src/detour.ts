import { LatLngLiteral } from "leaflet"
import { Shape, ShapePoint } from "./schedule"

export interface DetourShape {
  coordinates: ShapePoint[]
  directions: {
    instruction: string
  }[]
}

export interface OriginalRoute {
  routeName: string
  routeDescription: string
  routeOrigin: string
  routeDirection: string
  shape: Shape
  center: LatLngLiteral
  zoom: number
}
