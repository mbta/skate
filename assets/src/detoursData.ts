import { array, Infer, number, string, type } from "superstruct"
import { DraftDetours } from "./models/detours"

export const DraftDetoursData = array(
  type({
    route: string(),
    direction: string(),
    name: string(),
    intersection: string(),
    active_since: number(),
  })
)
export type DraftDetoursData = Infer<typeof DraftDetoursData>

export const draftDetoursFromData = (
  draftDetoursData: DraftDetoursData
): DraftDetours =>
  draftDetoursData.map((draft) => ({
    route: draft.route,
    direction: draft.direction,
    name: draft.name,
    intersection: draft.intersection,
    activeSince: draft.active_since,
  }))
