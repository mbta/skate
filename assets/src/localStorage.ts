export const saveState = (stateKey: string, state: object) => {
  const serializedState = JSON.stringify(state)
  localStorage.setItem(stateKey, serializedState)
}

export const loadState = (stateKey: string): object | undefined => {
  const serializedState = localStorage.getItem(stateKey)

  if (serializedState === null) {
    return undefined
  }

  return JSON.parse(serializedState)
}
