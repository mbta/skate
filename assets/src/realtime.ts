import {
  BlockId,
  DirectionId,
  RouteId,
  RoutePatternId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule"

import { Crowding } from "./models/crowding"
import { DetourId } from "./models/detoursList"

export interface BlockWaiver {
  startTime: Date
  endTime: Date
  causeId: number
  causeDescription: string
  remark: string | null
}

export interface DataDiscrepancy {
  attribute: string
  sources: DataDiscrepancySource[]
}

export interface DataDiscrepancySource {
  id: SourceId
  value: string | null
}

export type EndOfTripType = "another_trip" | "swing_off" | "pull_back"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  routePatternId: RoutePatternId | null
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  incomingTripDirectionId: DirectionId | null
  layoverDepartureTime: number | null
  scheduledTimepointStatus: VehicleTimepointStatus | null
  scheduledLogonTime: number | null
  routeStatus: RouteStatus
  blockWaivers: BlockWaiver[]
  currentPieceStartPlace: string | null
  currentPieceFirstRoute: RouteId | null
}

export type NotificationId = string

// A Notification's `content` is populated with an Elixir struct and the
// `__struct__` field is serialized so that the type can be discerned on the
// frontend.
export enum NotificationType {
  BridgeMovement = "Elixir.Notifications.Db.BridgeMovement",
  BlockWaiver = "Elixir.Notifications.Db.BlockWaiver",
  Detour = "Elixir.Notifications.Db.Detour",
}

export interface BridgeLoweredNotification {
  $type: NotificationType.BridgeMovement
  status: "lowered"
}

export interface BridgeRaisedNotification {
  $type: NotificationType.BridgeMovement
  status: "raised"
  loweringTime: Date
}

export type BridgeNotification =
  | BridgeLoweredNotification
  | BridgeRaisedNotification

export type BlockWaiverNotification = {
  $type: NotificationType.BlockWaiver
  createdAt: Date
  reason: BlockWaiverReason
  routeIds: RouteId[]
  runIds: RunId[]
  tripIds: TripId[]
  operatorId: string | null
  operatorName: string | null
  routeIdAtCreation: string | null
  startTime: Date
  endTime: Date | null
}

export enum DetourNotificationStatus {
  Activated = "activated",
  Deactivated = "deactivated",
}

export type DetourNotification = {
  $type: NotificationType.Detour
  status: DetourNotificationStatus
  detourId: DetourId
  headsign: string
  route: string
  direction: string
  origin: string
}

export type NotificationContentTypes =
  | BridgeNotification
  | BlockWaiverNotification
  | DetourNotification

export interface Notification<
  TNotification extends NotificationContentTypes = NotificationContentTypes
> {
  id: NotificationId
  createdAt: Date
  state: NotificationState
  content: TNotification
}

export function isBlockWaiverNotification(
  notification: Notification
): notification is Notification<BlockWaiverNotification> {
  return notification.content.$type === NotificationType.BlockWaiver
}
export type BlockWaiverReason =
  | "other"
  | "accident"
  | "adjusted"
  | "disabled"
  | "diverted"
  | "manpower"
  | "operator_error"
  | "traffic"

export type NotificationState = "unread" | "read" | "deleted"

export type SourceId = string

export type RouteStatus = "on_route" | "laying_over" | "pulling_out"

export type RunId = string

export interface ByRunId<T> {
  [runId: string]: T
}

export type TrainVehicleId = string

export interface TrainVehicle {
  id: TrainVehicleId
  latitude: number
  longitude: number
  bearing: number
}

export interface Vehicle {
  id: VehicleId
  label: string | null
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId | null
  routeId: RouteId | null
  routePatternId: RoutePatternId | null
  tripId: TripId | null
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string | null
  operatorFirstName: string | null
  operatorLastName: string | null
  operatorLogonTime: Date | null
  overloadOffset?: number
  bearing?: number
  blockId: BlockId | null
  previousVehicleId: string | null
  scheduleAdherenceSecs: number | null
  incomingTripDirectionId: DirectionId | null
  isShuttle: boolean
  isOverload: boolean
  isOffCourse: boolean
  isRevenue: boolean | null
  layoverDepartureTime: number | null
  pullbackPlaceName: string | null
  dataDiscrepancies: DataDiscrepancy[]
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  routeStatus: RouteStatus
  endOfTripType: EndOfTripType
  blockWaivers: BlockWaiver[]
  crowding: Crowding | null
}

export interface VehicleInScheduledService extends Vehicle {
  directionId: DirectionId
  routeId: RouteId
  blockId: BlockId
}

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId | null
  stopName: string | null
}

export interface VehicleScheduledLocation {
  routeId: RouteId
  routePatternId: RoutePatternId | null
  directionId: DirectionId
  tripId: TripId
  runId: RunId
  timeSinceTripStartTime: number
  headsign: string | null
  viaVariant: ViaVariant | null
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}
