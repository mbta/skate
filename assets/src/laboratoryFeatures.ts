import appData from "./appData"

const featureIsEnabled = (key: string): boolean => {
  const data = appData()

  if (data === undefined || data.laboratoryFeatures === undefined) {
    return false
  }

  const features = JSON.parse(data.laboratoryFeatures)
  if (features[key] === undefined) {
    return false
  }

  return features[key]
}

export default featureIsEnabled
