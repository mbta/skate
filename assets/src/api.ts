import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
