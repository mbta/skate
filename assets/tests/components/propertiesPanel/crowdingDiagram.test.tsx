import React from "react"
import renderer from "react-test-renderer"
import CrowdingDiagram from "../../../src/components/propertiesPanel/crowdingDiagram"
import { Crowding } from "../../../src/realtime"

describe("CrowdingDiagram", () => {
  test("renders nothing if route isn't considered reliable", () => {
    const tree = renderer.create(<CrowdingDiagram crowding={null} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders correctly for an untrusted APC", () => {
    const untrustedApcCrowding = {
      load: null,
      capacity: null,
      occupancyStatus: "NO_DATA",
      occupancyPercentage: null,
    } as Crowding

    const tree = renderer
      .create(<CrowdingDiagram crowding={untrustedApcCrowding} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders correctly for an empty bus", () => {
    const emptyCrowding: Crowding = {
      load: 0,
      capacity: 18,
      occupancyStatus: "EMPTY",
      occupancyPercentage: 0.0,
    }

    const tree = renderer
      .create(<CrowdingDiagram crowding={emptyCrowding} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders correctly for an uncrowded bus", () => {
    const uncrowded: Crowding = {
      load: 1,
      capacity: 20,
      occupancyStatus: "MANY_SEATS_AVAILABLE",
      occupancyPercentage: 0.05,
    }

    const tree = renderer
      .create(<CrowdingDiagram crowding={uncrowded} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders correctly for a somewhat crowded bus", () => {
    const somewhatCrowded: Crowding = {
      load: 10,
      capacity: 20,
      occupancyStatus: "FEW_SEATS_AVAILABLE",
      occupancyPercentage: 0.5,
    }

    const tree = renderer
      .create(<CrowdingDiagram crowding={somewhatCrowded} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders correctly for a crowded bus", () => {
    const crowded: Crowding = {
      load: 45,
      capacity: 30,
      occupancyStatus: "FULL",
      occupancyPercentage: 1.5,
    }

    const tree = renderer
      .create(<CrowdingDiagram crowding={crowded} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
