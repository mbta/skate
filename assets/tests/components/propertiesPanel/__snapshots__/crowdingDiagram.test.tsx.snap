// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrowdingDiagram renders correctly for a crowded bus 1`] = `
<div
  className="m-crowding-diagram"
>
  <div
    className="m-crowding-diagram__properties"
  >
    <span
      className="m-properties-list__property-label"
    >
      Riders onboard
    </span>
    <div
      className="mock-tippy"
    >
      <div
        className="mock-tippy-content"
      >
        <div>
          Riders are estimated using Automated 
          <br />
           Passenger Counters (APCs).
        </div>
      </div>
      <span
        className="m-crowding-diagram__tooltip-anchor"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </div>
    <br />
    45 riders
     / 
    30
     seats
    <br />
    <span
      className="m-crowding-diagram__status-description m-crowding-diagram__status-description--crowded"
    >
      Crowded
    </span>
  </div>
  <div
    className="m-crowding-diagram__crowding-icon-wrapper"
  >
    <span
      className="m-crowding-diagram__crowding-icon m-crowding-diagram__crowding-icon--crowded"
      dangerouslySetInnerHTML={
        Object {
          "__html": "SVG",
        }
      }
    />
  </div>
</div>
`;

exports[`CrowdingDiagram renders correctly for a somewhat crowded bus 1`] = `
<div
  className="m-crowding-diagram"
>
  <div
    className="m-crowding-diagram__properties"
  >
    <span
      className="m-properties-list__property-label"
    >
      Riders onboard
    </span>
    <div
      className="mock-tippy"
    >
      <div
        className="mock-tippy-content"
      >
        <div>
          Riders are estimated using Automated 
          <br />
           Passenger Counters (APCs).
        </div>
      </div>
      <span
        className="m-crowding-diagram__tooltip-anchor"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </div>
    <br />
    10 riders
     / 
    20
     seats
    <br />
    <span
      className="m-crowding-diagram__status-description m-crowding-diagram__status-description--some-crowding"
    >
      Some crowding
    </span>
  </div>
  <div
    className="m-crowding-diagram__crowding-icon-wrapper"
  >
    <span
      className="m-crowding-diagram__crowding-icon m-crowding-diagram__crowding-icon--some-crowding"
      dangerouslySetInnerHTML={
        Object {
          "__html": "SVG",
        }
      }
    />
  </div>
</div>
`;

exports[`CrowdingDiagram renders correctly for an empty bus 1`] = `
<div
  className="m-crowding-diagram"
>
  <div
    className="m-crowding-diagram__properties"
  >
    <span
      className="m-properties-list__property-label"
    >
      Riders onboard
    </span>
    <div
      className="mock-tippy"
    >
      <div
        className="mock-tippy-content"
      >
        <div>
          Riders are estimated using Automated 
          <br />
           Passenger Counters (APCs).
        </div>
      </div>
      <span
        className="m-crowding-diagram__tooltip-anchor"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </div>
    <br />
    0 riders
     / 
    18
     seats
    <br />
    <span
      className="m-crowding-diagram__status-description m-crowding-diagram__status-description--empty"
    >
      Empty
    </span>
  </div>
  <div
    className="m-crowding-diagram__crowding-icon-wrapper"
  >
    <span
      className="m-crowding-diagram__crowding-icon m-crowding-diagram__crowding-icon--empty"
      dangerouslySetInnerHTML={
        Object {
          "__html": "SVG",
        }
      }
    />
  </div>
</div>
`;

exports[`CrowdingDiagram renders correctly for an uncrowded bus 1`] = `
<div
  className="m-crowding-diagram"
>
  <div
    className="m-crowding-diagram__properties"
  >
    <span
      className="m-properties-list__property-label"
    >
      Riders onboard
    </span>
    <div
      className="mock-tippy"
    >
      <div
        className="mock-tippy-content"
      >
        <div>
          Riders are estimated using Automated 
          <br />
           Passenger Counters (APCs).
        </div>
      </div>
      <span
        className="m-crowding-diagram__tooltip-anchor"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </div>
    <br />
    1 rider
     / 
    20
     seats
    <br />
    <span
      className="m-crowding-diagram__status-description m-crowding-diagram__status-description--not-crowded"
    >
      Not crowded
    </span>
  </div>
  <div
    className="m-crowding-diagram__crowding-icon-wrapper"
  >
    <span
      className="m-crowding-diagram__crowding-icon m-crowding-diagram__crowding-icon--not-crowded"
      dangerouslySetInnerHTML={
        Object {
          "__html": "SVG",
        }
      }
    />
  </div>
</div>
`;

exports[`CrowdingDiagram renders correctly for an untrusted APC 1`] = `
<div
  className="m-crowding-diagram"
>
  <div
    className="m-crowding-diagram__properties"
  >
    <span
      className="m-properties-list__property-label"
    >
      Riders onboard
    </span>
    <div
      className="mock-tippy"
    >
      <div
        className="mock-tippy-content"
      >
        <div>
          Riders are estimated using Automated 
          <br />
           Passenger Counters (APCs).
        </div>
      </div>
      <span
        className="m-crowding-diagram__tooltip-anchor"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </div>
    <br />
    No data available
  </div>
  <div
    className="m-crowding-diagram__crowding-icon-wrapper"
  >
    <span
      className="m-crowding-diagram__crowding-icon m-crowding-diagram__crowding-icon--no-data"
      dangerouslySetInnerHTML={
        Object {
          "__html": "SVG",
        }
      }
    />
  </div>
</div>
`;

exports[`CrowdingDiagram renders nothing if route isn't considered reliable 1`] = `null`;
